Add revalidation check to Output Manager service

philipr-za opened the pull request:
## Description
Outputs can become marked as invalid by the Output Manager service when they do not appear in the base nodes blockchain. Generally this means that they were re-orged out or have been spent by another copy of the wallet. However they can be reported as invalid incorrectly in some case like, for example, if the Base Node is not fully synced.

This PR adds in an addition check that the Output Manager does on startup that will see if any of the invalid outputs has become valid again. If it has that output will changed back into a spendable output.

## How Has This Been Tested?
The test of the UTXO query has been updated to include checking this revalidation behaviour.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

sdbondi commented on base_layer/wallet/src/output_manager_service/service.rs:
```diff
@@ -413,70 +482,81 @@ where
+        utxo_query_timeout_futures: &mut FuturesUnordered<BoxFuture<'static, u64>>,
+    ) -> Result<u64, OutputManagerError>
+    {
+        let invalid_output_hashes: Vec<Vec<u8>> = self
+            .db
+            .get_invalid_outputs()
+            .await?
+            .iter()
+            .map(|uo| uo.hash.clone())
+            .collect();
```
Unnecessary clones
```suggestion
        let invalid_output_hashes: Vec<_> = self
            .db
            .get_invalid_outputs()
            .await?
            .into_iter()
            .map(|uo| uo.hash)
            .collect();
```

sdbondi reviewed (commented):
LGTM - What other triggers make sense for querying invalid outputs? Connecting to a new base node with a higher chain tip? (falls under #1805) 

One comment to resolve then I'll approve

philipr-za committed:
Add revalidation check to Output Manager service

Outputs can become marked as invalid by the Output Manager service when they do not appear in the base nodes blockchain. Generally this means that they were re-orged out or have been spent by another copy of the wallet. However they can be reported as invalid incorrectly in some case like, for example, if the Base Node is not fully synced.

This PR adds in an addition check that the Output Manager does on startup that will see if any of the invalid outputs has become valid again. If it has that output will changed back into a spendable output.
